llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -288,6 +292,86 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorInserts(
   return true;
 }
 
+/// Determines if it is safe to combine vextract by checking the uses of DstReg,
+/// specifically for a pattern involving TRUNC followed by EXT.
+bool AIE2PreLegalizerCombinerImpl::isTruncExtToS20Sequence(
+    Register DstReg, bool SignVal, unsigned SrcEltSize) const {
+  // Returns the single non-debug use of a register with a specific opcode
+  // and destination size.
+  auto GetOneUseWithOpcode =
+      [&](const Register Reg, const unsigned OpcodeToCheck,
+          const unsigned DstSize) -> std::optional<MachineInstr *> {
+    if (MRI.hasOneNonDBGUser(Reg)) {
+      MachineInstr &Use = *MRI.use_nodbg_instructions(Reg).begin();
+      if (Use.getOpcode() == OpcodeToCheck) {
+        const LLT DstRegTy = MRI.getType(Use.getOperand(0).getReg());
+        if (DstRegTy.getSizeInBits() == DstSize)
+          return &Use;
+      }
+    }
+    return std::nullopt;
+  };
+  auto Trunc = GetOneUseWithOpcode(DstReg, TargetOpcode::G_TRUNC, SrcEltSize);
+  if (!Trunc)
+    return false;
+
+  const MachineInstr *TruncMI = *Trunc;
+  const unsigned ExtOpcode =
+      SignVal ? TargetOpcode::G_SEXT : TargetOpcode::G_ZEXT;
+  const Register UseDstReg = TruncMI->getOperand(0).getReg();
+  return GetOneUseWithOpcode(UseDstReg, ExtOpcode, 20).has_value();
+}
+
+/// \returns true if it is possible to combine the below sequence of MIRs
+/// From : %3:_(s32) = G_INTRINSIC
+///         intrinsic(@llvm.aie2.vextract.elem[8/16].I512), %2(<32 x s16>),
+///         %0(s32), %1(s32)
+///        %4:_(s16) = G_TRUNC %3(s32)
+///        %5:_(s20) = G_SEXT %4(s16)
+/// To :   %9:_(s20) = G_AIE_SEXT_EXTRACT_VECTOR_ELT %2(<32 x s16>), %0(s32)
+///        %10:_(s20) = G_ASSERT_[S/Z]EXT %9, 16
+///        %4:_(s16) = G_TRUNC %10(s20)
+///        %5:_(s20) = G_[S/Z]EXT %4(s16)
```

Comment:
Do we need to change the return types? I would expect that we only need to add a `%10:_(s32) = G_ASSERT_[S/Z]EXT %9, 16` and keep the rest intact with the new `sext(trunc x)` combiner you added previously.